### PR TITLE
fix(changelog): gitRef check

### PR DESCRIPTION
**Pull request by atorrescogollo**
## Changes

For some dependencies, (in my case `kubernetes/autoscaler`), they don't have a `Changelog.md` and the releases are mixed inside the repo:
```
...
https://github.com/kubernetes/autoscaler/releases/tag/cluster-autoscaler-chart-9.18.1
https://github.com/kubernetes/autoscaler/releases/tag/cluster-autoscaler-1.24.0
...
```

I'm using this datasource configuration:
```
datasource=github-releases depName=kubernetes/autoscaler extractVersion=^cluster-autoscaler-(?<version>[0-9\.]*)$
```

And I'm receiving this `res` variable when trying to get the Changelog:

```
// source-github.ts -> getChangeLogJSON
res = {
  project: {
    apiBaseUrl: 'https://api.github.com/',
    baseUrl: 'https://github.com/',
    type: 'github',
    repository: 'kubernetes/autoscaler',
    sourceUrl: 'https://github.com/kubernetes/autoscaler',
    sourceDirectory: undefined,
    depName: 'kubernetes/autoscaler'
  },
  versions: [
    {
      version: '1.24.0',
      gitRef: 'cluster-autoscaler-1.24.0',
      date: '2022-05-16T14:57:43.000Z',
      changes: [],
      compare: [Object]
    },
    ...
  ]
}
```

As you can see, `version` tag is different from `gitRef` and this comparison won't ever match for this use-case:
```nodejs
matchedRelease = releases.find(
      (r) => r.tag === version || r.tag === `v${version}`
```

We need to check also if matches the `gitRef` so that it can download the Changelog from releases notes.

## Documentation (please check one with an [x])

- [ ] I have updated the documentation, or
- [x] No documentation update is required

## How I've tested my work (please tick one)

I have verified these changes via:

- [ ] Code inspection only, or
- [ ] Newly added/modified unit tests, or
- [x] No unit tests but ran on a real repository, or
- [ ] Both unit tests + ran on a real repository

### CODE REVIEW

**Review by viceice — COMMENTED** [review]
please validate that it works for gitlab too

**Comment by atorrescogollo** [comment]
For some reason, https://github.com/renovatebot/renovate/pull/15653 along with this PR seems to break `Release Notes` message in the PR body:
* Rebase from main (commit https://github.com/renovatebot/renovate/commit/2e957baed962d65cb8e40136edc142af6014ad95 applied) (https://github.com/renovatebot/renovate/pull/15653):
image

* Reverting commit https://github.com/renovatebot/renovate/commit/2e957baed962d65cb8e40136edc142af6014ad95 (https://github.com/renovatebot/renovate/pull/15653):
image

**Comment by rarkins** [comment]
> For some reason, #15653 along with this PR seems to break `Release Notes` message in the PR body:

That PR also refactored how tags are fetched for release notes. Please see if you can work out how to make the changes, otherwise @zharinov can help

**Comment by zharinov** [comment]
> For some reason, https://github.com/renovatebot/renovate/pull/15653 along with this PR seems to break Release Notes message

Found the reason, fix is ready

**Comment by atorrescogollo** [comment]
Thanks, I think I found another way of doing this thing but also using `config.extractVersion`, which I think makes more sense. I will be publishing the change tomorrow for review

**Comment by atorrescogollo** [comment]
Already rebased from main, fixed tests, added a new test for `config.extractVersion` and run tests locally

File: lib/workers/repository/update/pr/changelog/release-notes.spec.ts [hunk 1]
```diff
@@ -94,9 +95,16 @@ describe('workers/repository/update/pr/changelog/release-notes', () => {
   describe('addReleaseNotes()', () => {
     it('returns input if invalid', async () => {
       const input = { a: 1 };
-      expect(await addReleaseNotes(input as never)).toEqual(input);
-      expect(await addReleaseNotes(null)).toBeNull();
-      expect(await addReleaseNotes({ versions: [] } as never)).toStrictEqual({
+      expect(
+        await addReleaseNotes(input as never, {} as BranchUpgradeConfig)
```
Comment (viceice) [inline lib/workers/repository/update/pr/changelog/release-notes.spec.ts:99]: i think those `input` objects should have a proper  type, you can use `partial<TYPE>({ a: 1})` helper function from test utils

File: lib/workers/repository/update/pr/changelog/release-notes.ts [hunk 2]
```diff
@@ -107,9 +111,20 @@ export async function getReleaseNotes(
...
+        r.tag === `v${gitRef}`
     );
   }
+  if (is.undefined(matchedRelease) && config.extractVersion) {
+    const extractVersionRegEx = regEx(config.extractVersion);
+    matchedRelease = releases.find((r) => {
+      const extractedVersion = extractVersionRegEx.exec(r.tag)?.groups?.version;
+      return version === extractedVersion;
+    });
+  }
```
Comment (viceice) [inline lib/workers/repository/update/pr/changelog/release-notes.ts:127]: This looks like a separate fix to me, so should be done in a separate pr

File: lib/workers/repository/update/pr/changelog/release-notes.spec.ts (hunk 1 above)
Comment (atorrescogollo) [reply lib/workers/repository/update/pr/changelog/release-notes.spec.ts:99]: This test is `returns input if invalid`, so I can't cast `{ a: 1}` to a `ChangeLogResult` object even partially. Actually, I didn't create this test case but it's changed due to the linting after adding a new parameter to the function

File: lib/workers/repository/update/pr/changelog/release-notes.ts (hunk 2 above)
Comment (atorrescogollo) [reply lib/workers/repository/update/pr/changelog/release-notes.ts]: Created https://github.com/renovatebot/renovate/pull/15962 and already reverted changes for extractVersion flow

**Review by viceice — APPROVED** [review]

**Comment by viceice** [comment]
CLA check hung 😞
